src/codegen.cpp

```diff
@@ -3140,6 +3140,16 @@ static void emit_assignment(jl_value_t *l, jl_value_t *r, jl_codectx_t *ctx)
                     emit_unbox(vtype, slot, slot.typ),
                     false, slot.typ, ctx);
         }
+        else if (slot.isboxed) {
+            // see if inference had a better type for the gensym than the expression (after inlining getfield on a Tuple)
+            jl_value_t *gensym_types = jl_lam_gensyms(ctx->ast);
+            if (jl_is_array(gensym_types)) {
+                jl_value_t *declType = jl_cellref(gensym_types, idx);
+                if (declType != slot.typ) {
```

Comment:
if inference is wrong, it's going to fail far worse than simply mismarking it here...